src/Features/Core/Portable/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs

```diff
@@ -181,9 +183,94 @@ private void AnalyzeSyntax(SyntaxNodeAnalysisContext context, INamedTypeSymbol e
+            {
+                case TBinaryExpressionSyntax binaryExpression:
+                    return TryAnalyzeBinaryExpressionCondition(
+                        syntaxFacts, binaryExpression, out conditionLeft, out conditionRight, out isEquals);
+
+                case TInvocationExpression invocation:
+                    return TryAnalyzeInvocationCondition(
+                        context, syntaxFacts, referenceEqualsMethod, invocation,
+                        out conditionLeft, out conditionRight, out isEquals);
+
+                default:
+                    return TryAnalyzePatternCondition(
+                        conditionNode, out conditionLeft, out conditionRight, out isEquals);
+            }
+        }
+
+        private bool TryAnalyzeBinaryExpressionCondition(
+            ISyntaxFactsService syntaxFacts, TBinaryExpressionSyntax condition,
+            out SyntaxNode conditionLeft, out SyntaxNode conditionRight, out bool isEquals)
+        {
+            isEquals = IsEquals(condition);
+            if (!isEquals && !IsNotEquals(condition))
+            {
+                conditionLeft = null;
+                conditionRight = null;
+                return false;
+            }
+            else
+            {
+                syntaxFacts.GetPartsOfBinaryExpression(condition, out conditionLeft, out conditionRight);
+                return true;
+            }
+        }
+
+        private static bool TryAnalyzeInvocationCondition(
+            SyntaxNodeAnalysisContext context, ISyntaxFactsService syntaxFacts, IMethodSymbol referenceEqualsMethod, TInvocationExpression invocation,
+            out SyntaxNode conditionLeft, out SyntaxNode conditionRight, out bool isEquals)
+        {
+            conditionLeft = null;
+            conditionRight = null;
+            isEquals = true;
+
+            var expression = syntaxFacts.GetExpressionOfInvocationExpression(invocation);
+            var nameNode = syntaxFacts.IsIdentifierName(expression)
+                ? expression
+                : syntaxFacts.IsSimpleMemberAccessExpression(expression)
+                    ? syntaxFacts.GetNameOfMemberAccessExpression(expression)
+                    : null;
+
+            if (!syntaxFacts.IsIdentifierName(nameNode))
+            {
+                return false;
+            }
+
+            syntaxFacts.GetNameAndArityOfSimpleName(nameNode, out var name, out _);
+            if (!syntaxFacts.StringComparer.Equals(name, nameof(ReferenceEquals)))
+            {
+                return false;
+            }
+
+            var arguments = syntaxFacts.GetArgumentsOfInvocationExpression(invocation);
+            if (arguments.Count != 2)
+            {
+                return false;
+            }
+
+            var semanticModel = context.SemanticModel;
+            var cancellationToken = context.CancellationToken;
+            var symbol = semanticModel.GetSymbolInfo(invocation, cancellationToken).Symbol;
```

Comment:
:bulb: `GetSymbolInfo` is quite expensive. Consider updating the code to place the syntax-based null checks (the calls to `IsNullLiteralExpression`) before the call to `GetSymbolInfo`. I'm not overly concerned since you do have the name equality check a few lines above and `ReferenceEquals` isn't a particularly common name, so if it's not easy to make the change feel free to just create a new issue to address this later.

Reply:
Note: this should be fine in practice IMO.  By the time we're checking what symbol this is, we've already ensure the shape is ```ReferenceEquals(e1, e2) ? ... : ...``` that's going to be syntactically rare enough that we'll only being doing the semantic check sparingly.

Reply:
It wasn't to difficult to shuffle this around so I changed it so that `GetSymbolInfo` is called last.